### PR TITLE
SONARPY-618 Introduce Ambiguous Symbol

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**

### CODE REVIEW

File: python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java [hunk 1]
```diff
@@ -132,13 +133,73 @@ public void visitFileInput(FileInput fileInput) {
     scopesByRootTree = new HashMap<>();
     fileInput.accept(new FirstPhaseVisitor());
     fileInput.accept(new SecondPhaseVisitor());
+    createAmbiguousSymbols();
```
Comment (pynicolas) [inline python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java]: I don't like changing symbols at this point. It seems very late. What are the risks that we end up with a graph of symbols which doesn't make sense? Aren't we changing symbols which are referenced by other symbols?

File: python-frontend/src/main/java/org/sonar/python/types/TypeShed.java [hunk 2]
```diff
@@ -66,25 +67,33 @@ private TypeShed() {
...
+    TypeAnnotation returnTypeAnnotation = functionDef.returnTypeAnnotation();
+    if (returnTypeAnnotation == null) {
+      return;
+    }
+    if (symbol.is(Symbol.Kind.FUNCTION)) {
+      FunctionSymbolImpl functionSymbol = (FunctionSymbolImpl) symbol;
+      functionSymbol.setDeclaredReturnType(InferredTypes.declaredType(returnTypeAnnotation, builtins));
+    } else if (symbol.is(Symbol.Kind.AMBIGUOUS)) {
+      AmbiguousSymbolImpl ambiguousSymbol = (AmbiguousSymbolImpl) symbol;
+      setDeclaredReturnType(ambiguousSymbol.getSymbol(functionDef.name()), functionDef);
```
Comment (pynicolas) [inline python-frontend/src/main/java/org/sonar/python/types/TypeShed.java]: I think we should find a way to do that differently, without needing `AmbiguousSymbolImpl` to store `symbolsByDeclarationTree`.

Comment (pynicolas) [reply python-frontend/src/main/java/org/sonar/python/types/TypeShed.java]: One idea:
Add a `symbol` method to FunctionDefImpl. This symbol could be different from the symbol attached to the name of the function in the case of an ambiguous symbol.

Comment (andrea-guarino-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/types/TypeShed.java]: Good idea, I will go for it! Thanks!

File: python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java [hunk 3]
```diff
@@ -132,13 +133,69 @@ public void visitFileInput(FileInput fileInput) {
...
+    }
+  }
+
+  private void createAmbiguousSymbols() {
+    for (Scope scope : scopesByRootTree.values()) {
+      Set<SymbolToUpdate> symbolsToUpdate = new HashSet<>();
+      for (Symbol symbol : scope.symbols()) {
+        if (symbol.kind() == Symbol.Kind.OTHER) {
+          List<Usage> bindingUsages = symbol.usages().stream().filter(Usage::isBindingUsage).collect(Collectors.toList());
+          if (bindingUsages.size() > 1) {
```
Comment (pynicolas) [inline python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java:160]: I think that this condition is not sufficient. We end up creating an ambiguous symbol in the following case and I suppose we should not:
```
x = 42
x = 43
```

Comment (andrea-guarino-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java:160]: Good point. I'll update the condition to also check the presence of a usage of kind `FUNC_DECLARATION` or `CLASS_DECLARATION`

**Pull request merged**